### PR TITLE
Completely redo the QM buildscript into Java

### PR DESCRIPTION
**Pull request by OroArmor**
I might make this pull request more extensive and change the entire buildscript to java

### CODE REVIEW

**Comment by i509VCB** [comment]
I'd probably go all in on a java buildscript using a build logic structure similar to qsl

**Comment by OroArmor** [comment]
yeah, thats what my initial thought was

File: buildSrc/build.gradle [hunk 1]
```diff
@@ -0,0 +1,40 @@
+plugins {
+	id 'java'
+	id 'java-gradle-plugin'
+}
+
+repositories {
+	mavenLocal()
```
Comment (IotaBread) [inline buildSrc/build.gradle]: `mavenLocal` should be the last repository

File: buildSrc/build.gradle [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	implementation("com.google.guava:guava:30.1.1-jre")
+	implementation("cuchaz:enigma-cli:${project.enigma_version}")
+	implementation("cuchaz:enigma:${project.enigma_version}")
+}
+
+gradlePlugin {
+	plugins {
+		mappingsLogic {
+			id = 'mappings-logic'
+			implementationClass = 'quilt.internal.MappingsPlugin'
```
Comment (IotaBread) [inline buildSrc/build.gradle:41]: Why `quilt.internal`?

File: buildSrc/gradle.properties [hunk 3]
```diff
@@ -0,0 +1,2 @@
+stitch_version=0.6.1
+enigma_version=1.3.4
```
Comment (IotaBread) [inline buildSrc/gradle.properties]: Why this? These versions are already in the root `gradle.properties`

File: buildSrc/build.gradle (hunk 2 above)
Comment (OroArmor) [reply buildSrc/build.gradle:41]: not sure. I think it was used in the qsl buildscript, and i went off of that

File: buildSrc/gradle.properties (hunk 3 above)
Comment (OroArmor) [reply buildSrc/gradle.properties]: I couldnt get it to pull from the root properties

File: buildSrc/build.gradle (hunk 1 above)
Comment (OroArmor) [reply buildSrc/build.gradle]: this will be excluded in the final pr

File: buildSrc/build.gradle (hunk 2 above)
Comment (i509VCB) [reply buildSrc/build.gradle:41]: build-logic is not exposed to the outside world, so the package name isn't a huge issue here

File: build.gradle [hunk 4]
```diff
@@ -91,510 +104,48 @@ dependencies {
 	decompileClasspath "net.fabricmc:cfr:${project.cfr_version}"
 	mappingPoetJar "org.quiltmc:mapping-poet:${project.mapping_poet_version}"
 	unpick "net.fabricmc.unpick:unpick-cli:${project.unpick_version}"
+	hashed "org.quiltmc:hashed-mojmap:${minecraft_version}${USE_SNAPSHOT_HASHES ? "-SNAPSHOT" : ""}"
 }
 
-def setupGroup = "jar setup"
-def mappingsGroup = "quilt-mappings"
-def buildMappingGroup = "mapping build"
 def mapJarGroup = "jar mapping"
```
Comment (IotaBread) [inline build.gradle]: `mapJarGroup` isn't used anywhere

File: build.gradle [hunk 5]
```diff
@@ -91,510 +104,48 @@ dependencies {
...
 def mergedFile = file("${minecraft_version}-merged.jar")
-def hashedMojmapJar = file("${minecraft_version}-hashed-mojmap.jar")
-def unpickedJar = file("${minecraft_version}-hashed-mojmap-unpicked.jar")
+def hashedMojmapJar = file("${minecraft_version}-hashed.jar")
+def unpickedJar = file("${minecraft_version}-hashed-unpicked.jar")
 def namedJar = file("${minecraft_version}-named.jar")
-def versionFile = new File(cacheFilesMinecraft, "${minecraft_version}.json")
-def clientJar = new File(cacheFilesMinecraft, "${minecraft_version}-client.jar")
-def serverJar = new File(cacheFilesMinecraft, "${minecraft_version}-server.jar")
 def libraries = new File(cacheFilesMinecraft, "libraries")
```
Comment (IotaBread) [inline build.gradle]: These can be changed by their `FileConstants` equivalent (you can get the instance with `project.getExtensions().getByType(MappingsExtension.class).getFileConstants()`)

File: buildSrc/src/main/java/quilt/internal/tasks/AbstractMappingsTask.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import quilt.internal.util.DownloadImmediate;
+
+public interface AbstractMappingsTask extends Task {
+
+    default DownloadImmediate.Builder startDownload() {
+        return new DownloadImmediate.Builder(this);
+    }
+
+    @SuppressWarnings("unchecked")
+    default <T extends Task> T getTaskFromName(String taskName) {
```
Comment (IotaBread) [inline buildSrc/src/main/java/quilt/internal/tasks/AbstractMappingsTask.java]: `getTaskByName`?

File: buildSrc/src/main/java/quilt/internal/tasks/AbstractMappingsTask.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+        return (T) getProject().getTasks().getByName(taskName);
+    }
+
+    default <T extends Task> T getTaskFromType(Class<T> taskClass) {
```
Comment (IotaBread) [inline buildSrc/src/main/java/quilt/internal/tasks/AbstractMappingsTask.java]: `getTaskByType`?

**Review by IotaBread — APPROVED** [review]

File: buildSrc/gradle.properties (hunk 3 above)
Comment (IotaBread) [reply buildSrc/gradle.properties]: We should change to Quilt Enigma

File: build.gradle [hunk 8]
```diff
@@ -904,69 +436,35 @@ publishing {
 			}
 			artifact sourcesJar
 			artifact javadocJar
-		}
 
+			pom.withXml {
+				def hashedDep = it.asNode().appendNode('dependencies').appendNode('dependency')
+				hashedDep.appendNode('groupId', 'org.quiltmc')
+				hashedDep.appendNode('artifactId', 'hashed')
```
Comment (IotaBread) [inline build.gradle:343]:
```suggestion
				hashedDep.appendNode('artifactId', 'hashed-mojmap')
```

**Review by IotaBread — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 8 above)
Comment (OroArmor) [reply build.gradle:343]: the namespace in tiny is hashed, not hashed-mojmap.

Comment (IotaBread) [reply build.gradle:343]: Isn't this supposed to point to an artifact?

Comment (OroArmor) [reply build.gradle:343]: true. I didn't think of that. @CheaterCodes do we want to either move the mapping namespace to hashed-mojmap, or move the artifact to hashed

Comment (IotaBread) [reply build.gradle:343]: I'd say moving the artifact is the best option

File: build.gradle [hunk 9]
```diff
@@ -91,548 +104,52 @@ dependencies {
 	decompileClasspath "net.fabricmc:cfr:${project.cfr_version}"
 	mappingPoetJar "org.quiltmc:mapping-poet:${project.mapping_poet_version}"
 	unpick "net.fabricmc.unpick:unpick-cli:${project.unpick_version}"
+	hashed "org.quiltmc:hashed-mojmap:${minecraft_version}${USE_SNAPSHOT_HASHES ? "-SNAPSHOT" : ""}"
```
Comment (IotaBread) [inline build.gradle]:
```suggestion
	hashed "org.quiltmc:hashed:${minecraft_version}${USE_SNAPSHOT_HASHES ? "-SNAPSHOT" : ""}"
```
Artifact was moved

File: build.gradle [hunk 10]
```diff
@@ -875,7 +372,8 @@ javadoc {
 		// lazy setting
 		options {
 			tagletPath configurations.mappingPoet.files.toList()
-			header mappingPoetJar.get().filter { it.name == 'javadoc_header.txt' }.singleFile.text.trim() // cannot include line breaks
+			header mappingPoetJar.get().filter { it.name == 'javadoc_header.txt' }.singleFile.text.trim()
+			// cannot include line breaks
```
Comment (IotaBread) [inline build.gradle:287]: This comment isn't in the right place?

File: build.gradle [hunk 11]
```diff
@@ -927,69 +425,35 @@ publishing {
 			}
 			artifact sourcesJar
 			artifact javadocJar
-		}
 
+			pom.withXml {
+				def hashedDep = it.asNode().appendNode('dependencies').appendNode('dependency')
+				hashedDep.appendNode('groupId', 'org.quiltmc')
+				hashedDep.appendNode('artifactId', 'hashed')
+				hashedDep.appendNode('version', configurations.hashed.getDependencies().iterator().next().getVersion())
```
Comment (IotaBread) [inline build.gradle]: While we still have Hashed in the snapshot maven, this should point to the resolved hashed version, ie: `21w41a-20211013.160757-1`

File: build.gradle [hunk 12]
```diff
@@ -927,69 +425,35 @@ publishing {
...
+			}
+		}
 	}
 	repositories {
-		if (ENV.MAVEN_URL) {
+		if (ENV.MAVEN_URL && !USE_SNAPSHOT_HASHES) {
```
Comment (IotaBread) [inline build.gradle]: This means QM will be published to the snapshot maven while we use snapshot hashed? I'd keep it in the release maven

File: buildSrc/src/main/java/quilt/internal/Constants.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package quilt.internal;
+
+public class Constants {
+    public static final String MINECRAFT_VERSION = "21w40a";
```
Comment (IotaBread) [inline buildSrc/src/main/java/quilt/internal/Constants.java]: Is there a way to get this from a properties file? So we don't have to change the Minecraft version in two different places (here and the buildscript)

File: buildSrc/src/main/java/quilt/internal/tasks/build/DownloadPerVersionMappingsTask.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    public static final String TASK_NAME = "downloadHashedMojmap";
+
+    @OutputFile
+    public File jarFile = new File(fileConstants.cacheFilesMinecraft, String.format("%s-%s.jar", Constants.MINECRAFT_VERSION, Constants.PER_VERSION_MAPPINGS_NAME));
+
+    @OutputFile
+    public File tinyFile = new File(fileConstants.cacheFilesMinecraft, String.format("%s-%s.tiny", Constants.MINECRAFT_VERSION, Constants.PER_VERSION_MAPPINGS_NAME));
+
+    public DownloadPerVersionMappingsTask() {
+        super(Constants.Groups.BUILD_MAPPINGS_GROUP);
```
Comment (IotaBread) [inline buildSrc/src/main/java/quilt/internal/tasks/build/DownloadPerVersionMappingsTask.java]: I'd say this fits better in the setup group

**Review by IotaBread — COMMENTED** [review]
Other than my comments, looks good to me

File: buildSrc/src/main/java/quilt/internal/Constants.java (hunk 13 above)
Comment (OroArmor) [reply buildSrc/src/main/java/quilt/internal/Constants.java]: Im planning on some way to configure this automatically. If i can do it, there will be an update mappings task, which will change this automatically

File: buildSrc/build.gradle [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+	implementation("net.fabricmc:stitch:${project.stitch_version}")
+	implementation('org.quiltmc:tiny-remapper:0.7.1')
+	implementation('org.quiltmc:launchermeta-parser:1.0.0')
+	implementation("com.google.guava:guava:30.1.1-jre")
+	implementation("cuchaz:enigma-cli:${project.enigma_version}")
+	implementation("cuchaz:enigma:${project.enigma_version}")
+	implementation("net.fabricmc.unpick:unpick:${project.unpick_version}")
+	implementation("net.fabricmc.unpick:unpick-format-utils:${project.unpick_version}")
+	implementation("net.fabricmc.unpick:unpick-cli:${project.unpick_version}")
+	implementation("net.fabricmc.filament:net.fabricmc.filament.gradle.plugin:0.3.3")
```
Comment (IotaBread) [inline buildSrc/build.gradle:34]:
```suggestion
	implementation("org.quiltmc:fabric-filament:0.3.3")
```
Should work, I'm not 100% sure

File: buildSrc/gradle.properties [hunk 16]
```diff
@@ -0,0 +1,12 @@
+# Done to increase the memory available to gradle.
+org.gradle.jvmargs=-Xmx1G
+
+enigma_version=1.4.1
```
Comment (IotaBread) [inline buildSrc/gradle.properties]:
```suggestion
enigma_version=1.0.4
```

**Review by IotaBread — APPROVED** [review]
Fabric removed `buildTinyWithMappings`, we might want to do that as well (see https://github.com/FabricMC/yarn/commit/66374095feeac0c898c9d1bc7eee2c7490ccb21e)

File: build.gradle (hunk 11 above)
Comment (IotaBread) [reply build.gradle]: This snippet should work
```groovy
import org.gradle.api.internal.artifacts.repositories.resolver.MavenUniqueSnapshotComponentIdentifier
// ...
def id = configurations.hashed.getResolvedConfiguration().getResolvedArtifacts().iterator().next().getId().getComponentIdentifier()
def version
if (id instanceof MavenUniqueSnapshotComponentIdentifier) {
	version = id.getVersion().toString().replace("SNAPSHOT", id.getTimestamp())
} else {
	version = id.getVersion().toString()
}
```

Slightly shorter version
```groovy
def id = configurations.hashed.getResolvedConfiguration().getResolvedArtifacts().iterator().next().getId().getComponentIdentifier()
def version = id.toString().replace("SNAPSHOT:", "")
version = version.substring(version.lastIndexOf(":") + 1)
```

**Review by Jamalam360 — APPROVED** [review]
looks good

**Review by IotaBread — APPROVED** [review]

**Review by LambdAurora — APPROVED** [review]

**Pull request merged**